runtime/tests/interpreter/container_mutation_test.go

```diff
@@ -650,6 +648,221 @@ func TestDictionaryMutation(t *testing.T) {
+			array.Get(inter, interpreter.ReturnEmptyLocationRange, 0),
+		)
+		assert.Equal(
+			t,
+			interpreter.NewStringValue("hello from bar"),
+			array.Get(inter, interpreter.ReturnEmptyLocationRange, 1),
+		)
+	})
+
+	t.Run("bound function mutation", func(t *testing.T) {
+		t.Parallel()
+
+		inter := parseCheckAndInterpret(t, `
+	       struct Foo {
+	           fun foo(): String {
+	               return "hello from foo"
+	           }
+	       }
+
+	       struct Bar {
+	           fun bar(): String {
+	               return "hello from bar"
+	           }
+	       }
+
+	       fun test(): [String] {
+	           let dict: {String: AnyStruct} = {}
+
+	           let a = Foo()
+	           let b = Bar()
+
+	           dict["foo"] = a.foo
+	           dict["bar"] = b.bar
+
+	           let callFoo = dict["foo"]! as! (():String)
+	           let callBar = dict["bar"]! as! (():String)
+
+	           return [callFoo(), callBar()]
+	       }
+	   `)
+
+		value, err := inter.Invoke("test")
+		require.NoError(t, err)
+
+		require.IsType(t, &interpreter.ArrayValue{}, value)
+		array := value.(*interpreter.ArrayValue)
+
+		require.Equal(t, 2, array.Count())
+		assert.Equal(
+			t,
+			interpreter.NewStringValue("hello from foo"),
+			array.Get(inter, interpreter.ReturnEmptyLocationRange, 0),
+		)
+		assert.Equal(
+			t,
+			interpreter.NewStringValue("hello from bar"),
+			array.Get(inter, interpreter.ReturnEmptyLocationRange, 1),
+		)
+	})
+
+	t.Run("invalid function mutation", func(t *testing.T) {
+		t.Parallel()
+
+		standardLibraryFunctions :=
+			stdlib.StandardLibraryFunctions{
+				stdlib.LogFunction,
+			}
+
+		valueDeclarations := standardLibraryFunctions.ToSemaValueDeclarations()
+		values := standardLibraryFunctions.ToInterpreterValueDeclarations()
+
+		inter, err := parseCheckAndInterpretWithOptions(t, `
+	           fun test() {
+	               let dict: {String: AnyStruct} = {} as {String: (():Void)}
+
+	               dict["log"] = log
+	           }
+	       `,
+			ParseCheckAndInterpretOptions{
+				CheckerOptions: []sema.Option{
+					sema.WithPredeclaredValues(valueDeclarations),
+				},
+				Options: []interpreter.Option{
+					interpreter.WithPredeclaredValues(values),
+				},
+			},
+		)
+
+		require.NoError(t, err)
+
+		_, err = inter.Invoke("test")
+		require.Error(t, err)
+
+		mutationError := &interpreter.ContainerMutationError{}
+		require.ErrorAs(t, err, mutationError)
+
+		require.IsType(t, &sema.OptionalType{}, mutationError.ExpectedType)
+		optionalType := mutationError.ExpectedType.(*sema.OptionalType)
+
+		require.IsType(t, &sema.FunctionType{}, optionalType.Type)
+		funcType := optionalType.Type.(*sema.FunctionType)
+
+		assert.Equal(t, sema.VoidType, funcType.ReturnTypeAnnotation.Type)
+		assert.Nil(t, funcType.ReceiverType)
+		assert.Empty(t, funcType.Parameters)
+	})
+
+	t.Run("valid function mutation", func(t *testing.T) {
+		t.Parallel()
+
+		inter := parseCheckAndInterpret(t, `
+            struct S {}
```

Comment:
```suggestion
              struct S {}
```

Reply:
Cleaned up the whole file in 6e064c5a77fffe70d3db7c2b9035062348b0ac92